text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,641 @@
+conventions requires some additions and clarifications in the specification,
+which are listed in this section.
+
+* [open-telemetry/opentelemetry-specification#454](https://github.com/open-telemetry/opentelemetry-specification/issues/454)
+  To instrument push-based "Receive" operations as described in this document,
+  it is necessary to add links to spans after those spans were created. The
+  reason for this is, that not all messages are present at the start of a
+  "Receive" operations, so links to related contexts cannot be added at the start
+  of the span.
+* [open-telemetry/opentelemetry-specification#2176](https://github.com/open-telemetry/opentelemetry-specification/issues/2176)
+  When consuming a message with no attached creation context as part of a
+  batch, it would still be useful to capture related message-specific
+  attributes as part of a link which points to an invalid context. However,
+  according to the specification links pointing to an invalid context may be
+  ignored. For consistently reporting message-specific attributes on links, links
+  to invalid contexts should be allowed and supported.
+* [open-telemetry/opentelemetry-specification#3172](https://github.com/open-telemetry/opentelemetry-specification/issues/3172)
+  Currently the specification is unclear about whether relationships between
+  producer and consumer spans can be modelled via links, the wording suggests
+  that it should be a parent/child relationship. The wording in the specification
+  needs to make it clear that this can be a link too.
+
+## Examples
+
+This section contains a list of examples illustrating the use of the
+conventions outlined above. Green boxes denote spans that are required to
+exist in order to conform to those conventions. Other boxes denote spans that
+are not required and covered by the conventions, but are hopefully helpful in
+understanding how messaging spans can be integrated into an overall trace flow.
+Solid arrows denote parent/child relationships, dotted arrows denote link
+relationships.
+
+### Push-based scenarios
+
+A producer creates and publishes a single message, the single message is delivered to a consumer:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  PM1[Publish m1]
+  end
+  subgraph CONSUMER
+  DM1[Deliver m1]
+  end
+  PM1-. link .->DM1;
+
+  classDef normal fill:green
+  class PM1,DM1 normal
+  linkStyle 0 color:green,stroke:green
+```
+
+When consuming a single message, the "Deliver" spans can be parented to the creation context:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  PM1[Publish m1]
+  end
+  subgraph CONSUMER
+  DM1[Deliver m1]
+  end
+  PM1-. link .->DM1;
+  PM1-- parent -->DM1;
+
+  classDef normal fill:green
+  class PM1,DM1 normal
+  linkStyle 0,1 color:green,stroke:green
+```
+
+It is recommended to add spans for settlement operations. Those spans can
+either be created manually or via auto-instrumentation:
```

Comment:
@carlosalberto about your points above:

> Example of auto Settlement span?

This diagram exemplifies settlement spans already, but it's not clear if it's "manual" or "auto" settlement. We agreed to modify the text above the diagram to make it more clear. Do you think that addresses it? Here's a preview of how it would look:

![image](https://github.com/open-telemetry/oteps/assets/5938087/03217de2-3b70-45d7-80e5-e577daeeb257)